HTML interface module: Port of the crew monitoring computer

nullquery opened the pull request:
~~**Warning: This PR depends on PR #9849 to be merged first.** Once that PR has been merged the changed files list will update.~~ - <B>This has been Merged</B>

This contents of this PR modify the crew monitoring computer to utilize the new html_interface module.

**Note: Partially reworked, please see my large comment below for additional information.**

Improvements:
- Updates the items and positions dynamically as players move (CPU profiler-tested with ~45 "fake" players).
- Some optimization of the way that it updates the HTML (in addition to the performance improvements by simply using the html_interface module).
- Patch: Don't update html_interface clients that have been inactive for more than 10 minutes.
- Patch: Attempt to speed up _renderContent some more (in case of multiple clients) by using spawn.
- Patch: Second argument for hiIsValidClient hook: reference to the current html_interface object (used by crew monitoring computer).
- Feature: procqueue singleton. You can use this to put proc executions on a queue. Used by the crew monitoring computer to queue the update "for the next tick" when humans move.
- Debug verb: Populate World. You can use this to quickly add sNPCs at various random locations in the game. To help simulate a live server. (Side-effect: sNPCs are also spawned in the AI upload chamber, so expect tasers/projectiles to be actively used as well.)

Screenshot:

![image](https://cloud.githubusercontent.com/assets/5695364/8391172/63ba8a14-1cb8-11e5-8399-bea036a26560.png)

Commit:
- Crew monitoring computer using the html_interface module
- Patch: Don't update html_interface clients that have been inactive for more than 10 minutes.
- Patch: Attempt to speed up _renderContent some more (in case of multiple clients) by using spawn.
- Patch: Second argument for hiIsValidClient hook: reference to the current html_interface object (used by crew monitoring computer).
- Feature: procqueue singleton. You can use this to put proc executions on a queue. Used by the crew monitoring computer to queue the update "for the next tick" when humans move.

Commit:
Merge pull request #5 from nullstation/master

Merge with upstream

Commit:
Fix for the "Populate world" verb.

Commit:
Merge branch 'hi-extended' of https://github.com/nullstation/-tg-station into hi-extended

Commit:
Changelog entry.

RemieRichards commented on code/controllers/subsystem/procqueue.dm:
```diff
@@ -0,0 +1,48 @@
+	src.queue.Insert(1, item)
+
+/datum/procqueue_item
+	var
```
please don't define variables like this.

var/ref
var/procname
var/list/args 
please.

nullquery replied:
According to the CONTRIBUTING.md file (https://github.com/tgstation/-tg-station/blob/eb5d9ee8bcf5efbc54a5c7d7fb9dd2c98ba97f43/CONTRIBUTING.md) this doesn't need to happen for variables defined at class-level.

I've gone ahead and changed it, but I'd like to know if this is something I need to take note of in future commits or if there's confusion here.

RemieRichards replied:
Reading the contributing.md file, specifically the "Specification" section shows the example of a variable defined in absolute paths, and then states you must write all byond code in absolute paths.

nullquery replied:
Ah, now I notice the difference. In objects new variables are prefixed with "var/" but are not typed as an absolute path. I assume this is what you meant?

RemieRichards commented on code/game/machinery/computer/crew.dm:
```diff
@@ -33,9 +33,17 @@
 		src.updateDialog()
 		return
 
+var/datum/singleton/crewmonitor/crewmonitor = new
```
singleton means nothing in byond.

you probably want:
```
var/global/datum/singleton/crewmonitor/crewmonitor
or var/static/.../crewmonitor
```

nullquery replied:
"Singleton" is a generic word used to reference a type of object that only exists once in the entire project (as is the case here).

Relevant wiki entry: https://en.wikipedia.org/wiki/Singleton_variable

It doesn't seem to be confined to specific languages. In fact, the word is used all the time in popular Java libraries even though Java doesn't support anything to enforce singletons.

The reason I added it in this case is so that a developer will immediately know that there will only be one object of that type, with the ulterior motive of making the code more readable.

nullquery replied:
To avoid confusion: I'm referring to the use of "singleton" in the path name. IIRC the use of "static" or "global" between "/var" and "/datum" is an instruction for BYOND and not part of the type path.

RemieRichards replied:
I know what a singelton is, and it means nothing in byond's type paths.
a global or static is byond's way of telling a developer "there is only one of these"

nullquery replied:
I added the word "singleton" to the type path, not the variable declaration. The idea is that it's immediately obvious that it's only used once, without even being aware there's a global variable with the same name.

I've changed the variable definition, but I'd rather not get rid of the word "singleton" in the type path. Is this ok?

RemieRichards replied:
I'd prefer you not keep it.
you claim it's for readability, it instead clutters the readability.

It's purpose is to explain that there is only ever one of these, in byond that's written as a /static or a /global in the variable declaration, when working with the class itself there's no real need to know that it's only going to exist once, we have plenty of functional singletons in the game that do not have /singleton in their typepath.

furthermore, adding more to a path increases compile time, something you said you wanted to reduce?

tl;dr I'd prefer you remove it, as it's against our standards, and is needless.

Commit:
Absolute paths with variable declarations.

Commit:
Fix for .swinset

Commit:
Added the word "global" to the variable declaration.

Commit:
Remove "singleton" from the crewmonitor datum.

Commit:
Merge pull request #6 from tgstation/master

Merge pull request #9849 from nullstation/master

d3athrow commented:
Would it be possible to have the crew monitor map with this system like nano's got?

nullquery commented:
If it works like I think it does (maps are pre-generated separate resources with dots placed on them according to the players' coordinates) then yes, it would be possible.

duncathan commented:
This map is fantastic. :+1:

Commit:
Merge pull request #7 from tgstation/master

Merge with upstream

Commit:
Merge pull request #8 from nullstation/master

Merge with upstream

nullquery commented:
It can be done better, so I'm looking into adding a button that switches to a minimap view (like NanoUI).

Commit:
Minimap support for the crew monitoring computer.

nullquery commented:
Done!

![image](https://cloud.githubusercontent.com/assets/5695364/8398139/080a08f2-1de4-11e5-9a20-7cd3a24591ac.png)

Changes since original message:
* Added minimap generation.
  * When the round starts a minimap is automatically generated for each z-level. This is a background process and does not cause any noticeable lag for the client.
  * The state of the map is cached. Minimaps are only (re)generated if the PNG files are missing or if the map has changed.
  * Limitation: Minimaps are NOT dynamically updated but instead represent the blueprints of the station at initial construction (i.e., at server start).
* When you open the crew monitoring computer, a hand-held crew monitoring scanner, or the crew monitoring window as AI you'll be confronted with a window similar to the screenshot above.
  * The left portion of the screen is dedicated to the minimap.
  * The right portion of the screen shows the old table view, with a few key differences:
    * To save space the information on health and the exact coordinates where the person is located are hidden. You can show these by hovering your mouse over the columns.
    * The information on health has been replaced with a health indicator. It goes from green to red to indicate health, with green being good and red being bad. (If full health information is not available you only have a single green/red color, otherwise it uses tones to indicate how severe the damage is based on the total damage.)
    * Scrolling will only scroll the table. The minimap does not move. This does mean you have to maintain a minimum window height of ~540 pixels to see the full map.
    * Hovering over a player name in the table highlights their position on the minimap. Hovering over a dot on the minimap highlights the player in the table.
    * **Not pictured:** The colors for the various ranks (and bold for heads) still applies. The color in the minimap is also set. This is not visible in my screenshots because none of the NPC's I spawned have valid ID's (and Captain has the default white color + bold).
    * As is the norm the contents of the window are updated dynamically. I've used the profiler and confirmed that CPU stays well within reasonable limits (<0.5). With 100 players (75 showing location, 25 only showing vitals) the CPU still stays within reasonable limits (not in the top 10, 15th position during stress test)
* Limitation: If a player walks out of range of a camera they will not show up on the minimap. Their coordinate position will still be visible by hovering over the location column, for non-visual manual tracking. (Reasoning: Makes it more difficult to track players who use the maintenance areas to hide, and prevents AI from being _too_ omnipresent.)
* Special features for the AI: clicking on the map takes you to the nearest camera. Clicking on a player (either the dot on the map or in the table) tracks the selected player.
* Limitations for AI:
  * The health indicator is always white for an AI player. You can still hover your mouse to see health information, but you can't see an overview of player health. (Reasoning: Prevents AI player from immediately reacting to other players being harmed, making rounds uninteresting.)

![image](https://cloud.githubusercontent.com/assets/5695364/8398219/684016ba-1de6-11e5-9c1c-b8c7ed3269d0.png)
**Highlighting either the dot or a table row causes this view.**

![image](https://cloud.githubusercontent.com/assets/5695364/8398231/950b6938-1de6-11e5-9950-287fd8691243.png)
**Looks like Ramon isn't feeling so good.**

![image](https://cloud.githubusercontent.com/assets/5695364/8398236/a711c6d6-1de6-11e5-821e-b0d6c1f690d9.png)
**Daniela jumped out of the airlock again. What a klutz!**

**API changes**

* html_interface objects now have a "closeAll" proc. Calling this will close windows on all clients. This is automatically used on all html_interface objects if a cinematic is triggered (mainly so people won't continue to see the minimap if the station blows up, but now it'll close all computers that use the new html_interface module I guess).
* Added a "callJavascript" function to execute Javascript functions. This is necessary for the clearing and updating of the minimap/table.
* Bug fix: The background of the Nanotrasen-style windows was not shown correctly in IE8.

Commit:
Merge pull request #9 from tgstation/master

Merge with upstream

Anonus commented:
:+1: Is cool, is a thing that's really nice.

Commit:
Merge pull request #10 from nullstation/master

Merge with upstream

Xhuis commented:
YES

Razharas commented:
Id still prefer the colour changed from white to something more noticable like green of orange or purple
Not easy to notice people in medbay and R&D

xxalpha commented:
This is awesome.

Summoners commented:
:+1:

Ricotez commented:
:+1: x  :100:

ike709 commented:
:+1: this is beautiful

freerealestate commented:
Wow.

This is really impressive, especially by SS13/BYOND standards.

>The health indicator is always white for an AI player. You can still hover
your mouse to see health information, but you can't see an overview of
player health. (Reasoning: Prevents AI player from immediately reacting to
other players being harmed, making rounds uninteresting.)

If I had to nitpick though, I don't really like this. It's an inorganic
limitation, and it's not like the AI can have the crew monitoring console
open and maximised throughout the entire round (unless the player has a
second monitor I guess, but most people don't).

On 29 June 2015 at 15:20, ike709 <notifications@github.com> wrote:

> [image: :+1:] this is beautiful
>
> —
> Reply to this email directly or view it on GitHub
> <https://github.com/tgstation/-tg-station/pull/10202#issuecomment-116409790>
> .
>

duncathan commented:
Having it was the general consensus in #coderbus. Miauw in particular wanted to ensure the AI wasn't getting buffed especially by this, IIRC.

nullquery commented:
I'm one of those people who has more than 1 monitor, so I could have the monitor open at all times. :grin:

What dunc said. It was deliberately done to prevent AI from becoming too OP. I don't think it distracts too much.

Commit:
Update changelog

nullquery commented:
**I've looked it over once more, and I think this one can be merged at your discretion.** I have no further work, unless there are style or design issues that haven't been brought to my attention yet.

Regarding the white color: Unless you have an alternative ready I don't know what to change this to. The colors you mentioned are already in use by other ranks. White is the default "department-less" color.

If you wish to tinker with the colors you can do so by modifying the "crew.js" file. Look for "#FFFFFF", there's a fallback to use the white color for the minimap.

Other than that I think this is mergeable for now. I've updated the changelog with the relevant changes.

Iamgoofball commented:
@nullquery Do you think you can make hovering over a person on the list make their icon a different color along with big so it pops out a bit more?

Invisty commented:
Fantastic feature, but the ever-improving ease of usability leads to further unbalanced gameplay, particularly the accessibility of crew tracking consoles.

I would hope that the handheld trackers retain some limitation in the form of a crappy UI.

nullquery commented:
@Iamgoofball Sure, just need to modify the CSS file. I'll look into it.

@Invisty Players who are not near a camera will be hidden anyway, which should mitigate that issue. But if it becomes a problem we can just introduce more limitations.

Commit:
- Dot and table hover is now more prominent when hovering.
- Added transition effects for supported browsers (IE10+)
- Default color changed from white to brown.
- Special color for Captain.

nullquery commented:
Screenshot of latest change:

![image](https://cloud.githubusercontent.com/assets/5695364/8425033/0cbf1eae-1f02-11e5-8d91-e4468a91b27b.png)

Unless there are any other issues this is can be merged at your discretion.

RemieRichards commented:
My only issue is the "totally should be turfs but aren't turfs" catwalks not appearing on the map OR transport pods not appearing on maps.

but that's not your fault.

If nobody complains by the end of the day I'm merging.

Chnkr commented:
Haven't looked at the code yet, but does this map update, even when you are no longer in range of the machine itself?

nullquery commented:
@Chnkr No, it uses my html_interface module, which has built-in checks to prevent this kind of behavior. This design is deliberate, so that other developers who copy/paste code don't end up making the same mistakes, and to keep the range checks in one place.

Commit:
Lattices/catwalks and transit tubes now appear on the minimap.

nullquery commented:
@RemieRichards Done! Catwalks and transport pods (I'm assuming you meant transit tubes??) will now appear on the minimap.

This PR can still be merged.

Iamgoofball commented:
I'd recommend genning windows and doors on the map too.

palpatine213 commented:
Don't think it really needs windows and doors, since the map isn't really going to be able to be zoomed in. Adding too much detail will clutter things up I'd feel.

xxalpha commented:
You don't really need more than turfs, the tubes are a nice addition, but more than that is clutter.

Iamgoofball commented:
I dunno, it just looks kinda weird without the tables and windows and doors.

palpatine213 commented:
Have you tried actually compiling the branch and testing it? the map is like half the size of the picture. You wouldn't be able to see anything.

optimumtact commented:
Can you not just resize the browse window it's in @palpatine213 ?

RemieRichards commented:
@nullquery wow, you actually added the tubes and catwalks.
that's amazing, thanks.

I said I'd merge this at the end of yesterday but I fell asleep.
No more issues have been raised in this time, so It's a go.

palpatine213 commented:
YAY!

nullquery commented:
@optimumtact The window can be resized but the minimap image does not size along with it. Instead the table gets more space.

Glad to see this has been merged. Looking forward to the next thing I can contribute.

RemieRichards commented:
@nullquery I think we're all looking forward to it.

Actionb commented:
Some minor things why I actually prefer the old look. Essentially it felt easier to read:
- you could quickly and at a single glance obtain crucial information ('Who is dying and from what damage type?') without having to hover over icons. Finding somebody in crit usually only required scanning for somebody with o2 loss, a matter of seconds. 

- it seemed less cluttered due to being sorted not by first name only, but primarily by colour coded professions with low level jobs being at the bottom including unknowns. Now we have a full rainbow of colours, with 'unknowns' somehow intermixed all over the place (maybe they are still sorted by their real name & profession?), diverting attention from the real information you want to gather.
Jobs like assistant and service have a brown colour, somewhat mixing in with supply and engineering. White was fine for those.

- AI's abilities to scan for dead or dying crewmembers or to just generally get a feel of how healthy the crew is, feels really gimped. At times, even dead people show up with a white health indicator instead of a red one (bug?) which means you basically HAVE to hover over every single icon to find any casualties. As an AI I usually had some spare time to order a cyborg to go to x , y to recover a body, simply because doing so didn't take much time at all. Now I barely do that, because hovering over icons for 30 seconds just to find simple information is not worthit.

I hope this small bits can help you making this neat addition even better.

nullquery commented:
Getting information on who is dying should be easier; the icon goes from green to red which is a better indicator than constantly refreshing a web page. Don't see what I can do about that without losing too much space. Maybe I could add a border around the dots on the map to make it easier, but that may just make it OP again.

Sorting should not have been affected by this change. I'll look into this, since it's meant to sort how it used to (could have been an oversight on my part).

AI deliberately sees only a white indicator and has to hover for more information because this addition would have made the AI too OP. If you have alternatives that everyone can agree on I'll make the change but otherwise it'll be difficult for me to push those changes through.

There's still an outstanding to-do which I mentioned in another issue. I'll look into both issues this Sunday at the latest. It shouldn't take me long to come up with a fix.

freerealestate commented:
The list should really be sorted based on total damage, rather than
alphabetically. When you're scanning the list, all you're interested in are
the wounded crewmembers. It'd be nice not have to scroll all the way down
each time to find them.

Some of the colour choices are a little odd too. A lot of civilian jobs
(all, I think) have a brown colour which would be more suited to supply
jobs, instead their job-themed colour (e.g. chaplain black, mime white,
assistants grey or clown pink) or at least the colour of the civilian
department (though this would be less preferable due to being unable to
tell the different jobs apart at a glimpse). Head of Personnel is orange
when they should be command dark blue. Cargo is orange when it should
probably be brown, while engineering is yellow when it should probably be
orange.
